working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+- *C* declares a private-named and potentially non-nullable instance variable with no initializer expressions. 
+- Or in **NNBD-legacy mode**,  *C* declares a private-named and final instance variable with no initializer expression.
+
+Otherwise, for each constructor *s* of the superclass, named `S` or `S.id`, satisfying all of the following:
+
+1. If *s* is a named constructor with name `S.id` then:
+   1.  *s* is accessible ( `id` is not a private identifier or *S* is not declared in a different library), and
+   2.  *C* does not declare a static member named *id*.
+2. The *s* constructor is a generative constructor.
+3. *C* does not declare an instance variable with no initializer expression and with the same basename as a required named parameter of *s*.
+4. Either *C* does not declare any instance variables or *s* does not have any optional positional parameters (required because we can't declare a constructor with both optional position parameters and named parameters).
+
+add a generative default constructor *c* to *C* with the corresponding name `C` or `C.id`, the following parameters:
+
+- For each non-private instance variable with no initializer expression declared by *C*, where `v` is the name of the instance variable, the constructor has a named initializing formal parameter `this.v` with no default value. The named parameter is *required* if the instance variable has a potentially non-nullable type.
+- For each named parameter of *s*, *c* has a corresponding parameter with the same name, type, optionality and default value (if optional), unless a named initializing formal parameter with the same name was declared above (in which case the parameter of *s* must be optional per requirement 4. above).
+- For each positional parameter of *s*, whether optional or not, $c$ has a corresponding positional parameter with the same position, type, optionality and default value (if optional). For documentation purposes, the parameter will have the same name as the corresponding positional parameter of *s* unless that conflicts with one of the named initializing formals added in the previous point. In that case, the name will be modified by adding a `$` and a positive decimal integer numeral with no leading zeros after the name, where the numeral represents the smallest positive (&ge; 1) integer sufficient to make the name not be the same as any of the named parameters of the constructor, not the same as the name of any later positional parameter of *s*, and not the same as any earlier positional parameter of *c*.
+
+and with an initializer list consisting of a single `super` invocation of either `super`, if *s* is named `S`, or `super.id`, if *s* is named `S.id`, with one argument for each parameter of *c* corresponding to a parameter of *s*, forwarding the value of the *c* parameter as an argument to the corresponding *s* parameter.
+
+The constructor *c* is `const` if *s* is `const` and *C* does not declare *any* instance variables.
```

Comment:
I don't use const enough to recall all the dark corners, but it is not possible to support a default const constructor with fields? Is the problem that we don't want to silently give the user a *const* constructor unless they ask for it?